tools/gui/gui/sof_controller_engine.py

```diff
@@ -0,0 +1,197 @@
+
+        volume_match = re.search(r"numid=(\d+),iface=MIXER,name='(.*Master GUI Playback Volume.*)'", controls_output)
+        if volume_match:
+            volume_control = volume_match.group(2)
+            print(f"Detected Volume Control: {volume_control}")
+        else:
+            print("Master GUI Playback Volume control not found.")
+            raise RuntimeError("Volume control not found.")
+
+        eq_match = re.search(r"numid=(\d+),iface=MIXER,name='EQIIR1\.0 eqiir_coef_1'", controls_output)
+        if eq_match:
+            eq_numid = eq_match.group(1)
+            print(f"Detected EQ numid: {eq_numid}")
+        else:
+            print("EQ control not found.")
+            raise RuntimeError("EQ control not found.")
+
+    except subprocess.CalledProcessError as e:
+        print(f"Failed to run device detection commands: {e}")
+        raise
+
+def scale_volume(user_volume):
+    normalized_volume = user_volume / 100.0
+    scaled_volume = 31 * (math.sqrt(normalized_volume))
+    return int(round(scaled_volume))
+
+def scan_for_files(directory_name: str, file_extension: str, extra_paths: list = None):
+    found_files = []
+    dir_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), directory_name)
+
+    if os.path.exists(dir_path):
+        found_files.extend([f for f in os.listdir(dir_path) if f.endswith(file_extension)])
+    else:
+        print(f"Error: The '{directory_name}' directory is missing. It should be located in the same folder as this script.")
+
+    if extra_paths:
+        for path in extra_paths:
+            if os.path.exists(path):
+                found_files.extend([f for f in os.listdir(path) if f.endswith(file_extension)])
+            else:
+                print(f"Warning: The directory '{path}' does not exist.")
+
+    return found_files
+
+def execute_command(command: str, data: int = 0, file_name: str = None):
+    if device_string is None or volume_control is None or eq_numid is None:
+        raise RuntimeError("Device not initialized. Call initialize_device() first.")
+
+    command_switch = {
+        'volume': lambda x: handle_volume(data),
+        'eq': lambda x: handle_eq(file_name),
+        'play': lambda x: handle_play(file_name),
+        'pause': lambda x: handle_pause(),
+        'record': lambda x: handle_record(start=data, filename=file_name)
+    }
+
+    command_function = command_switch.get(command, lambda x: handle_unknown_command(data))
+    command_function(data)
+
+def handle_volume(data: int):
+    amixer_command = f"amixer -D{device_string} cset name='{volume_control}' {data}"
+    try:
+        subprocess.run(amixer_command, shell=True, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+    except subprocess.CalledProcessError as e:
+        print(f"Failed to set volume: {e}")
+
+def handle_eq(eq_file_name: str):
+    ctl_command = f"./sof-ctl -D{device_string} -n {eq_numid} -s {eq_file_name}"
```

Comment:
the -c with control name as text as shown by amixer controls is more robust than number value.  E.g. `-c name='DMIC0 Capture IIR Eq'` . The numbers change if topology is modified.

Reply:
I definitely agree with that, however one functionality of the controller engine is that it automatically detects the numid for these basic components. Check out the initialize_device() method in the controller engine, specifically lines 37-43 and 45-51.

Because of this autodetection, I think that using the explicit name leads to a higher chance of error overall.

Reply:
OK, I didn't know there's such feature. It sounds like the numids are then fine.